<a href="https://colab.research.google.com/github/nickcanoy/Masters_in_DataScience/blob/main/CANOY_Apriori_FPGrowth_ECLAT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

RODENICK CANOY | Data Mining

Discussion 7.3 | Apriori, FPGrowth and ECLAT Coding

Problem Setup (Transactions + Threshold Interpretation)
Requirement: “A database has five transactions”
We will encode the transactions as lists/sets.
From the image:

T100 = {M, O, N, K, E, Y}
T200 = {D, O, N, K, E, Y}
T300 = {M, A, K, E}
T400 = {M, U, C, K, Y}
T500 = {C, O, O, K, I, E} → treat as {C, O, K, I, E} (duplicates don’t change presence)

Requirement: “Let min sup = 60% and min conf = 80%”


With 5 transactions, support threshold count is:
0.60*5=30.60 \times 5 = 30.60*5=3
So any frequent itemset must appear in at least 3 transactions.

In [ ]:
#Install + Imports
!pip -q install mlxtend

import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, fpgrowth, association_rules

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


In [ ]:
#Encode Transactions

transactions = [
    ['M','O','N','K','E','Y'],      # T100
    ['D','O','N','K','E','Y'],      # T200
    ['M','A','K','E'],              # T300
    ['M','U','C','K','Y'],          # T400
    ['C','O','K','I','E']           # T500 (duplicate O removed)
]

te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)
df = pd.DataFrame(te_ary, columns=te.columns_)
df

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

,A,C,D,E,I,K,M,N,O,U,Y
0,False,False,False,True,False,True,True,True,True,False,True
1,False,False,True,True,False,True,False,True,True,False,True
2,True,False,False,True,False,True,True,False,False,False,False
3,False,True,False,False,False,True,True,False,False,True,True
4,False,True,False,True,True,True,False,False,True,False,False


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [ ]:
#Apriori Frequent Itemsets (min_support=0.60)

min_sup = 0.60

freq_ap = apriori(df, min_support=min_sup, use_colnames=True)
freq_ap['support_count'] = (freq_ap['support'] * len(transactions)).round().astype(int)
freq_ap.sort_values(['support','itemsets'], ascending=[False, True]).reset_index(drop=True)

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

,support,itemsets,support_count
0,1.0,(K),5
1,0.8,(E),4
2,0.8,"(E, K)",4
3,0.6,(M),3
4,0.6,(O),3
5,0.6,(Y),3
6,0.6,"(O, E)",3
7,0.6,"(M, K)",3
8,0.6,"(O, K)",3
9,0.6,"(Y, K)",3


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


In [ ]:
#Apriori Association Rules (min_confidence=0.80)

min_conf = 0.80

rules_ap = association_rules(freq_ap, metric="confidence", min_threshold=min_conf)

# add support counts for readability
rules_ap['support_count'] = (rules_ap['support'] * len(transactions)).round().astype(int)
rules_ap['antecedent_support_count'] = (rules_ap['antecedent support'] * len(transactions)).round().astype(int)
rules_ap['consequent_support_count'] = (rules_ap['consequent support'] * len(transactions)).round().astype(int)

rules_ap.sort_values(['confidence','lift','support'], ascending=[False, False, False]).reset_index(drop=True)

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski,support_count,antecedent_support_count,consequent_support_count
0,(O),(E),0.6,0.8,0.6,1.0,1.25,1.0,0.12,inf,0.5,0.75,1.0,0.875,3,3,4
1,"(O, K)",(E),0.6,0.8,0.6,1.0,1.25,1.0,0.12,inf,0.5,0.75,1.0,0.875,3,3,4
2,(O),"(E, K)",0.6,0.8,0.6,1.0,1.25,1.0,0.12,inf,0.5,0.75,1.0,0.875,3,3,4
3,(E),(K),0.8,1.0,0.8,1.0,1.00,1.0,0.00,inf,0.0,0.80,0.0,0.900,4,4,5
4,(M),(K),0.6,1.0,0.6,1.0,1.00,1.0,0.00,inf,0.0,0.60,0.0,0.800,3,3,5
5,(O),(K),0.6,1.0,0.6,1.0,1.00,1.0,0.00,inf,0.0,0.60,0.0,0.800,3,3,5
6,(Y),(K),0.6,1.0,0.6,1.0,1.00,1.0,0.00,inf,0.0,0.60,0.0,0.800,3,3,5
7,"(O, E)",(K),0.6,1.0,0.6,1.0,1.00,1.0,0.00,inf,0.0,0.60,0.0,0.800,3,3,5
8,(K),(E),1.0,0.8,0.8,0.8,1.00,1.0,0.00,1.0,0.0,0.80,0.0,0.900,4,5,4


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


Interpretations:


M → K has confidence 1.0 but lift 1.0

Because K is always present (support(K)=1.0)
Meaning: even without M, K is guaranteed; so M doesn’t “increase” K.
Interpretation: “Whenever M occurs, K occurs”—true, but not very informative because K is universal.



O → E has confidence 1.0 and lift 1.25

support(E)=0.80, confidence=1.0
lift = 1 / 0.80 = 1.25
Interpretation: When O appears, E becomes 25% more likely than baseline.



K → E has confidence 0.80 and lift 1.0

baseline support(E)=0.80
so knowing K doesn’t change probability of E (K always occurs).
Interpretation: Rule passes confidence but is weak by lift.



O → KE has confidence 1.0 and lift 1.25

support(KE)=0.80; confidence=1.0
Interpretation: Every time O appears, it always comes with both K and E in this dataset.

In [ ]:
#FP-Growth Frequent Itemsets
freq_fp = fpgrowth(df, min_support=min_sup, use_colnames=True)
freq_fp['support_count'] = (freq_fp['support'] * len(transactions)).round().astype(int)
freq_fp.sort_values(['support','itemsets'], ascending=[False, True]).reset_index(drop=True)

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

,support,itemsets,support_count
0,1.0,(K),5
1,0.8,(E),4
2,0.8,"(E, K)",4
3,0.6,(Y),3
4,0.6,(M),3
5,0.6,(O),3
6,0.6,"(Y, K)",3
7,0.6,"(M, K)",3
8,0.6,"(O, E)",3
9,0.6,"(O, K)",3


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [ ]:
#FP-Growth Association Rules

rules_fp = association_rules(freq_fp, metric="confidence", min_threshold=min_conf)

rules_fp['support_count'] = (rules_fp['support'] * len(transactions)).round().astype(int)
rules_fp.sort_values(['confidence','lift','support'], ascending=[False, False, False]).reset_index(drop=True)

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski,support_count
0,(O),(E),0.6,0.8,0.6,1.0,1.25,1.0,0.12,inf,0.5,0.75,1.0,0.875,3
1,"(O, K)",(E),0.6,0.8,0.6,1.0,1.25,1.0,0.12,inf,0.5,0.75,1.0,0.875,3
2,(O),"(E, K)",0.6,0.8,0.6,1.0,1.25,1.0,0.12,inf,0.5,0.75,1.0,0.875,3
3,(E),(K),0.8,1.0,0.8,1.0,1.00,1.0,0.00,inf,0.0,0.80,0.0,0.900,4
4,(Y),(K),0.6,1.0,0.6,1.0,1.00,1.0,0.00,inf,0.0,0.60,0.0,0.800,3
5,(M),(K),0.6,1.0,0.6,1.0,1.00,1.0,0.00,inf,0.0,0.60,0.0,0.800,3
6,(O),(K),0.6,1.0,0.6,1.0,1.00,1.0,0.00,inf,0.0,0.60,0.0,0.800,3
7,"(O, E)",(K),0.6,1.0,0.6,1.0,1.00,1.0,0.00,inf,0.0,0.60,0.0,0.800,3
8,(K),(E),1.0,0.8,0.8,0.8,1.00,1.0,0.00,1.0,0.0,0.80,0.0,0.900,4


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [ ]:
#ECLAT (Vertical Format)

from itertools import combinations

tids = list(range(1, len(transactions)+1))  # 1..5
trans_sets = [set(t) for t in transactions]

# Build vertical representation: item -> set of TIDs
vertical = {}
for tid, items in zip(tids, trans_sets):
    for it in items:
        vertical.setdefault(it, set()).add(tid)

min_count = int(min_sup * len(transactions))  # 3

def eclat(vertical_db, min_count):
    # frequent itemsets stored as {frozenset(items): tidset}
    frequent = {}

    # Start with 1-item candidates
    items = sorted(vertical_db.items(), key=lambda x: x[0])
    level = [(frozenset([it]), tidset) for it, tidset in items if len(tidset) >= min_count]

    k = 1
    while level:
        next_level = []
        # record all itemsets in this level
        for itemset, tidset in level:
            frequent[itemset] = tidset

        # generate candidates by joining
        for i in range(len(level)):
            for j in range(i+1, len(level)):
                a, tid_a = level[i]
                b, tid_b = level[j]
                union = a | b
                if len(union) == k + 1:
                    inter = tid_a & tid_b
                    if len(inter) >= min_count:
                        next_level.append((union, inter))

        # remove duplicates
        unique = {}
        for itemset, tidset in next_level:
            unique[itemset] = tidset
        level = [(k, v) for k, v in unique.items()]
        k += 1

    return frequent

freq_eclat = eclat(vertical, min_count)

# Convert to DataFrame
rows = []
for itemset, tidset in freq_eclat.items():
    rows.append({
        "itemsets": set(itemset),
        "support_count": len(tidset),
        "support": len(tidset)/len(transactions),
        "tids": sorted(list(tidset))
    })
df_eclat = pd.DataFrame(rows).sort_values(['support','support_count'], ascending=False).reset_index(drop=True)
df_eclat

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

,itemsets,support_count,support,tids
0,{K},5,1.0,"[1, 2, 3, 4, 5]"
1,{E},4,0.8,"[1, 2, 3, 5]"
2,"{E, K}",4,0.8,"[1, 2, 3, 5]"
3,{M},3,0.6,"[1, 3, 4]"
4,{O},3,0.6,"[1, 2, 5]"
5,{Y},3,0.6,"[1, 2, 4]"
6,"{O, E}",3,0.6,"[1, 2, 5]"
7,"{M, K}",3,0.6,"[1, 3, 4]"
8,"{O, K}",3,0.6,"[1, 2, 5]"
9,"{Y, K}",3,0.6,"[1, 2, 4]"


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

**Summary of Final Answers**

All Frequent Itemsets (min_sup=0.60)

1-itemsets: {K}, {E}, {M}, {O}, {Y}
2-itemsets: {K,E}, {M,K}, {O,K}, {O,E}, {K,Y}
3-itemsets: {O,K,E}

Strong Association Rules (min_conf=0.80)
Most important / interpretable ones:

O → E (conf=1.00, lift=1.25) → “If O is bought, E is always bought; E becomes 25% more likely than baseline.”

O → K (conf=1.00, lift=1.00) → always together but K is universal so not insightful

M → K (conf=1.00, lift=1.00) → same reasoning

Y → K (conf=1.00, lift=1.00)

K → E (conf=0.80, lift=1.00) → meets threshold but not informative

O → {K,E} (conf=1.00, lift=1.25) → “O implies both K and E”

{O,K} → E (conf=1.00, lift=1.25)

{O,E} → K (conf=1.00, lift=1.00)